jdbc-core/src/main/java/com/salesforce/datacloud/jdbc/core/accessor/impl/DataCloudArray.java

```diff
@@ -94,17 +127,15 @@ public void free() {
         // no-op
     }
 
-    private static Object getArrayNoBoundCheck(ValueVector dataVector, long start, long count) {
-        Object[] result = new Object[LargeMemoryUtil.checkedCastToInt(count)];
-        for (int i = 0; i < count; i++) {
-            result[i] = dataVector.getObject(LargeMemoryUtil.checkedCastToInt(start + i));
-        }
-        return result;
-    }
-
     private void checkBoundaries(long index, int count) {
-        if (index < 0 || index + count > this.startOffset + this.valuesCount) {
-            throw new ArrayIndexOutOfBoundsException();
+        // Check against actual stored data size using 1-based JDBC indexing
+        if (index < 1 || index > this.data.length) {
+            throw new ArrayIndexOutOfBoundsException("Index " + index + " is out of bounds for array size "
+                    + this.data.length + " (JDBC arrays use 1-based indexing)");
+        }
+        if (index + count - 1 > this.data.length) {
+            throw new ArrayIndexOutOfBoundsException("Index " + index + " + count " + count + " exceeds array size "
+                    + this.data.length + " (JDBC arrays use 1-based indexing)");
```

Comment:
I think the following formulation should be simpler as it follows the normal 0 based bound check logic
```suggestion
        long dataIndex = index - 1;
        // Check against actual stored data size using 1-based JDBC indexing
        if (dataIndex < 0 || dataIndex >= this.data.length) {
            throw new ArrayIndexOutOfBoundsException("Index " + index + " is out of bounds for array size "
                    + this.data.length + " (JDBC arrays use 1-based indexing)");
        }
        if ((dataIndex + count) >= this.data.length) {
            throw new ArrayIndexOutOfBoundsException("Index " + index + " + count " + count + " exceeds array size "
                    + this.data.length + " (JDBC arrays use 1-based indexing)");
```

Reply:
Modified the logic, this is simpler. Thank you, Moritz.